File: ClientFX/src/main/java/Controller/HomeController.java
package Controller;

import DTO.EventDTO;
import DTO.ParticipantDTO;
import contestUtils.IContestServices;
import example.model.Event;
import example.model.Participant;
import example.model.User;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;

import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import contestUtils.IMainObserver;

public class HomeController extends AnchorPane implements IMainObserver {
    private IContestServices server;
    private User currentUser;
    //private EventImplementationService eventService;
    //private ParticipantService participantService;
    private Stage currentStage;
    //private Properties properties;

    @FXML
    private ComboBox<Event> eventComboBox;

    @FXML
    private Label usernameLabel;

    @FXML
    private Label searchMessageLabel;

    @FXML
    private TableView<EventDTO> eventTable;

    @FXML
    private TableColumn<EventDTO, String> eventStyle;

    @FXML
    private TableColumn<EventDTO, Integer> eventDistance;

    @FXML
    private TableColumn<EventDTO, Integer> eventParticipantsCount;

    @FXML
    private TableView<Participant> participantTable;

    @FXML
    private TableColumn<Participant, String> participantName;

    @FXML
    private TableColumn<Participant, Integer> participantAge;

    @FXML
    private TableView<ParticipantDTO> searchResultsTable;

    @FXML
    private TableColumn<ParticipantDTO, String> searchNameColumn;

    @FXML
    private TableColumn<ParticipantDTO, Integer> searchAgeColumn;

    @FXML
    private TableColumn<ParticipantDTO, Integer> searchEventCountColumn;

    @FXML
    private VBox searchResultsContainer;

    @FXML
    private void onCloseSearchResults() {
        searchResultsContainer.setVisible(false);
    }

    @FXML
    protected void onSearchClicked() {
        Event selectedEvent = eventComboBox.getValue();
        if (selectedEvent == null) {
            showSearchMessage("Please select an event first!", true);
            return;
        }

        CompletableFuture.runAsync(() -> {
            try {
                Collection<ParticipantDTO> results = server.getParticipantsForEventWithCount(selectedEvent.getId());

                if (results == null || results.isEmpty()) {
                    Platform.runLater(() -> {
                        showSearchMessage("No participants found for this event.", true);
                        searchResultsContainer.setVisible(false);
                    });
                } else {
                    Platform.runLater(() -> {
                        updateSearchResults(results);
                        showSearchResults();
                    });
                }
            } catch (Exception e) {
                Platform.runLater(() -> {
                    showAlert("Search Error", "An error occurred: " + e.getMessage());
                    searchResultsContainer.setVisible(false);
                });
            }
        });
    }


    private void updateSearchResults(Collection<ParticipantDTO> results) {
        if (searchResultsTable != null) {
            CompletableFuture.runAsync(() -> {
                System.out.println("Search result size" + results.size());
                searchResultsTable.getItems().clear();
                searchResultsTable.getItems().addAll(results);
            });
        } else {
            System.err.println("searchResultsTable is null!");
        }
    }

    private void showSearchResults() {
        if (searchResultsContainer != null && searchResultsTable != null) {
            CompletableFuture.runAsync(() -> {
                searchResultsTable.setPlaceholder(new Label("No results found"));
                searchResultsContainer.setVisible(true);
                searchResultsTable.setVisible(true);
                searchMessageLabel.setVisible(false);
            });
        } else {
            System.err.println("Search results components not initialized!");
        }
    }

    @FXML
    protected void onLogoutClicked() {
        try{
            server.logout(currentUser,this);
        } catch (Exception e) {
            showAlert("Logout error ", e.getMessage());
        }
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(getClass().getResource("/views/main-view.fxml"));
            Scene scene = new Scene(fxmlLoader.load());
            MainController controller = fxmlLoader.getController();
            controller.init(server, currentStage);
            currentStage.setScene(scene);
            currentStage.show();
        } catch (IOException ioException) {
            System.out.println(ioException.getMessage());
        }
    }

    @FXML
    protected void onParticipantButtonClicked() {
        try {
            FXMLLoader loader = new FXMLLoader(getClass().getResource("/views/new-participant-view.fxml"));
            Parent root = loader.load();
            NewParticipantController controller = loader.getController();
            controller.init(server, currentUser, new Stage());

            Stage stage = new Stage();
            stage.setScene(new Scene(root));
            stage.setTitle("Add New Participant");
            stage.showAndWait();
            initialiseParticipantsTable();
        } catch (IOException e) {
            showAlert("Error", "Cannot open participant form: " + e.getMessage());
        }
    }

    @FXML
    protected void onNewEntryClicked() throws Exception {
        try {
            FXMLLoader loader = new FXMLLoader(getClass().getResource("/views/event-view.fxml"));
            Parent root = loader.load();
            EventEntriesController controller = loader.getController();
            controller.init(server, new Stage(), currentUser, this);

            Stage stage = new Stage();
            stage.setScene(new Scene(root));
            stage.setTitle("New Event Registration");
            stage.initOwner(currentStage);
            stage.showAndWait();
        } catch (IOException e) {
            showAlert("Error", "Failed to open new entry form: " + e.getMessage());
        }
        initialiseEventTable();
    }

    public void init(IContestServices server, User currentUser, Stage currentStage) {
        try {
            currentStage.setTitle("Swiming Contest");
            this.server = server;
            this.currentUser = currentUser;
            this.currentStage = currentStage;

            usernameLabel.setText(" (" + currentUser.getUserName() + ")");

            searchNameColumn.setCellValueFactory(new PropertyValueFactory<>("name"));
            searchAgeColumn.setCellValueFactory(new PropertyValueFactory<>("age"));
            searchEventCountColumn.setCellValueFactory(new PropertyValueFactory<>("eventCount"));
            searchResultsContainer.setVisible(false);

            initializeEventComboBox();

            Platform.runLater(() -> {
                try {
                    initializeEventComboBox();
                    initialiseParticipantsTable();
                    initialiseEventTable();
                } catch (Exception e) {
                    showAlert("Init error", "Eroare la inițializare tabele: " + e.getMessage());
                }
            });

        } catch (Exception e) {
            showAlert("An error occurred in init (HomeController): ", e.getMessage());
        }
    }


    private void initializeEventComboBox() {
        CompletableFuture.runAsync(() -> {
            try {
                var events = server.findAllEvents();
                Platform.runLater(() -> {
                    eventComboBox.getItems().clear();
                    eventComboBox.getItems().addAll(events);
                });
            } catch (Exception e) {
                Platform.runLater(() -> showAlert("Error", "Could not load events: " + e.getMessage()));
            }
        });
    }


    private void initialiseParticipantsTable()  {
        CompletableFuture.runAsync(() -> {
            try {
                participantTable.getItems().clear();
                //participantTable.setPlaceholder(new Label("No Participants"));
                participantName.setCellValueFactory(new PropertyValueFactory<>("name"));
                participantAge.setCellValueFactory(new PropertyValueFactory<>("age"));
                Collection<Participant> participants = server.findAllParticipants();
                Platform.runLater(() -> {
                    participantTable.getItems().addAll(participants);

                });
            } catch (Exception e) {
                Platform.runLater(() -> showAlert("Error", "Failed to load participants: " + e.getMessage()));
            }
        });
    }

    private void initialiseEventTable(){
        CompletableFuture.runAsync(() -> {
            try {
                eventTable.getItems().clear();
                eventTable.setPlaceholder(new Label("No Events"));
                eventStyle.setCellValueFactory(new PropertyValueFactory<>("style"));
                eventDistance.setCellValueFactory(new PropertyValueFactory<>("distance"));
                eventParticipantsCount.setCellValueFactory(new PropertyValueFactory<>("participantsCount"));
                Collection<EventDTO> events = server.getEventsWithParticipantsCount();
                Platform.runLater(() -> eventTable.getItems().addAll(events));
            } catch (Exception e) {
                Platform.runLater(() -> showAlert("Error", "Failed to load events: " + e.getMessage()));
            }
        });
    }

    private void showAlert(String title, String message) {
        Alert alert = new Alert(Alert.AlertType.ERROR);
        alert.setTitle(title);
        alert.setHeaderText(null);
        alert.setContentText(message);
        alert.showAndWait();
    }

    private void showSearchMessage(String message, boolean isError) {
        if (searchMessageLabel != null) {
            searchMessageLabel.setText(message);
            searchMessageLabel.setStyle(isError ? "-fx-text-fill: #a80c0c;" : "-fx-text-fill: #006400;");
            searchMessageLabel.setVisible(true);
        }
        if (searchResultsContainer != null) {
            searchResultsContainer.setVisible(false);
        }
    }

    @Override
    public void participantAdded(Participant participant) {
        Platform.runLater(() -> participantTable.getItems().add(participant));
    }

    @Override
    public void newRegistration(List<EventDTO> events) {
        Platform.runLater(() -> {
            eventTable.getItems().clear();
            eventTable.getItems().addAll(events);
        });
    }

    @Override
    public void eventAdded(EventDTO event) {
        Platform.runLater(() -> eventTable.getItems().add(event));
    }
}

File: java-app/Client/Controller/HomeController.cs
using mpp_proiect_csharp_DianaGliga11.Model;
using mpp_proiect_csharp_DianaGliga11.Model.DTO;
using Service;
using log4net;


namespace Controller
{
    public class HomeController : Form, IMainObserver
    {
        //private IDictionary<string, string> properties;
        private User _currentUser;
        private readonly IContestServices server;
        private static readonly ILog Log = LogManager.GetLogger(typeof(MainController));

        private ComboBox _eventComboBox;
        private Label _usernameLabel;
        private Label _searchMessageLabel;
        private DataGridView _eventTable;
        private DataGridView _participantTable;
        private DataGridView _searchResultsTable;
        private Panel _searchResultsContainer;
        private Button _btnCloseSearchResults;
        private Button _btnSearch;
        private Button _btnLogout;
        private Button _btnAddParticipant;
        private Button _btnNewEntry;
        
        public HomeController(IContestServices server)
        {
            this.server = server;
        }


        private void InitializeComponents()
        {
            this.Text = "Swimming Contest";
            this.Size = new Size(1000, 700);
            this.StartPosition = FormStartPosition.CenterScreen;

            _usernameLabel = new Label
            {
                Text = $"User: {_currentUser.UserName}",
                Location = new Point(20, 20),
                AutoSize = true
            };

            _eventComboBox = new ComboBox
            {
                Location = new Point(20, 50),
                Width = 200,
                DropDownStyle = ComboBoxStyle.DropDownList
            };

            _btnSearch = new Button
            {
                Text = "Search",
                Location = new Point(280, 50),
                Width = 80,
                Height = 30
            };
            _btnSearch.Click += OnSearchClicked;

            _searchMessageLabel = new Label
            {
                Location = new Point(390, 55),
                AutoSize = true,
                Visible = false
            };

            _eventTable = new DataGridView
            {
                Location = new Point(20, 90),
                Width = 450,
                Height = 200,
                ReadOnly = true,
                AllowUserToAddRows = false
            };
            _eventTable.Columns.Add("Style", "Style");
            _eventTable.Columns.Add("Distance", "Distance");
            _eventTable.Columns.Add("ParticipantsCount", "Participants Count");

            _participantTable = new DataGridView
            {
                Location = new Point(20, 310),
                Width = 450,
                Height = 200,
                ReadOnly = true,
                AllowUserToAddRows = false
            };
            _participantTable.Columns.Add("Name", "Name");
            _participantTable.Columns.Add("Age", "Age");

            _searchResultsContainer = new Panel
            {
                Location = new Point(500, 90),
                Size = new Size(450, 200),
                BorderStyle = BorderStyle.FixedSingle,
                Visible = false
            };

            _searchResultsTable = new DataGridView
            {
                Dock = DockStyle.Fill,
                ReadOnly = true,
                AllowUserToAddRows = false
            };
            _searchResultsTable.Columns.Add("Name", "Name");
            _searchResultsTable.Columns.Add("Age", "Age");
            _searchResultsTable.Columns.Add("EventsCount", "Events Count");
            _searchResultsContainer.Controls.Add(_searchResultsTable);

            _btnCloseSearchResults = new Button
            {
                Text = "Close",
                Height = 30,
                Dock = DockStyle.Bottom
            };
            _btnCloseSearchResults.Click += OnCloseSearchResults;
            _searchResultsContainer.Controls.Add(_btnCloseSearchResults);

            _btnAddParticipant = new Button
            {
                Text = "Add Participant",
                Location = new Point(20, 530),
                Width = 150,
                Height = 30
            };
            _btnAddParticipant.Click += OnParticipantButtonClicked;

            _btnNewEntry = new Button
            {
                Text = "New Event Entry",
                Location = new Point(180, 530),
                Width = 150,
                Height = 30
            };
            _btnNewEntry.Click += OnNewEntryClicked;

            _btnLogout = new Button
            {
                Text = "Logout",
                Location = new Point(800, 20),
                Width = 100,
                Height = 30
            };
            _btnLogout.Click += OnLogoutClicked;

            this.Controls.Add(_usernameLabel);
            this.Controls.Add(_eventComboBox);
            this.Controls.Add(_btnSearch);
            this.Controls.Add(_searchMessageLabel);
            this.Controls.Add(_eventTable);
            this.Controls.Add(_participantTable);
            this.Controls.Add(_searchResultsContainer);
            this.Controls.Add(_btnAddParticipant);
            this.Controls.Add(_btnNewEntry);
            this.Controls.Add(_btnLogout);
        }

        private void LoadEventComboBox()
        {
            UpdateUI(() =>
            {
                var events = server.GetAllEvents();
                if (events == null || !events.Any())
                {
                    MessageBox.Show("No events found.");
                    return;
                }

                _eventComboBox.DataSource = events.ToList();
                _eventComboBox.SelectedIndex = -1;

            });
        }


        public void UpdateUI(Action action)
        {
            if (InvokeRequired)
            {
                BeginInvoke(action);
            }
            else
            {
                action();
            }
        }

        private void LoadParticipants()
        {
            UpdateUI(() =>
            {
                _participantTable.Rows.Clear();
                foreach (var participant in server.GetAllParticipants())
                {
                    Log.Debug($"Loading participant: {participant.Name}");
                    _participantTable.Rows.Add(participant.Name, participant.Age, participant.Id);
                }
            });
        }

        private void LoadEvents()
        {
            if (InvokeRequired)
            {
                Invoke(new Action(LoadEvents));
                return;
            }

            _eventTable.Rows.Clear();
            var events = server.GetEventsWithParticipantsCount();
            foreach (var ev in events)
            {
                _eventTable.Rows.Add(ev.style, ev.distance, ev.participantsCount);
            }
        }

        private void OnSearchClicked(object sender, EventArgs e)
        {
            var selectedEvent = _eventComboBox.SelectedItem as Event;
            Log.Info($"Selected item: {selectedEvent}");
            if (selectedEvent == null)
            {
                ShowSearchMessage("Please select an event first!", true);
                return;
            }

            try
            {
                var results = server.GetParticipantsForEventWithCount(selectedEvent.Id);
                if (results == null || !results.Any())
                {
                    ShowSearchMessage("No participants found for this event.", true);
                    _searchResultsContainer.Visible = false;
                }
                else
                {
                    UpdateSearchResults(results);
                    _searchResultsContainer.Visible = true;
                }
            }
            catch (Exception ex)
            {
                ShowAlert("Search Error", ex.Message);
                _searchResultsContainer.Visible = false;
            }
        }

        private void UpdateSearchResults(IEnumerable<ParticipantDTO> results)
        {
            _searchResultsTable.Rows.Clear();
            foreach (var result in results)
            {
                _searchResultsTable.Rows.Add(result.name, result.age, result.eventCount);
            }
        }
        
        private void ShowSearchMessage(string message, bool isError)
        {
            _searchMessageLabel.Text = message;
            _searchMessageLabel.ForeColor = isError ? Color.DarkRed : Color.DarkGreen;
            _searchMessageLabel.Visible = true;
        }

        private void OnCloseSearchResults(object sender, EventArgs e)
        {
            _searchResultsContainer.Visible = false;
        }

        private void OnLogoutClicked(object sender, EventArgs e)
        {
            var loginForm = new MainController(server);
            loginForm.Show();
            this.Close();
        }

        private void OnParticipantButtonClicked(object sender, EventArgs e)
        {
            var allEvents = server.GetAllEvents();
            var allParticipants = server.GetAllParticipants();
            using (var newParticipantForm = new NewParticipantController(server, this, null, allEvents, allParticipants))
            {
                if (newParticipantForm.ShowDialog() == DialogResult.OK)
                {
                    LoadParticipants();
                }

            }
        }

        private void OnNewEntryClicked(object sender, EventArgs e)
        {
            var allEvents = server.GetAllEvents();
            var allParticipants = server.GetAllParticipants();
            using (var eventEntriesForm = new EventEntriesController(server, this, null, allEvents, allParticipants))
            {
                if (eventEntriesForm.ShowDialog() == DialogResult.OK)
                {
                    LoadEvents();
                }
            }
        }

        private void ShowAlert(string title, string message)
        {
            MessageBox.Show(message, title, MessageBoxButtons.OK, MessageBoxIcon.Error);
        }

        public void ParticipantAdded(Participant participant)
        {
            UpdateUI(() =>
            {
                _participantTable.Rows.Add(participant.Name, participant.Age); 
            });
        }

        public void EventEvntriesAdded(List<EventDTO> events)
        {
            UpdateUI(() => {
                _eventTable.Rows.Clear();
                foreach (var ev in events)
                {
                    _eventTable.Rows.Add(ev.style, ev.distance, ev.participantsCount);
                }
            });
        }

        public void SetLoggedInUser(User user)
        {
            this._currentUser = user;
            InitializeComponents();  

            _usernameLabel.Text = $"Logged in as: {_currentUser.UserName}";
            LoadEventComboBox();
            LoadParticipants();
            LoadEvents();
        }
    }
}
